File: src/rchain/embeddings.rs
use std::env;
use std::error::Error;

use reqwest::blocking::Client;
use serde_json::json;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EmbeddingProvider {
    Openai,
    Fireworks,
}

impl EmbeddingProvider {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Openai => "openai",
            Self::Fireworks => "fireworks",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ChunkStrategy {
    Paragraph,
    Sentence,
    Token,
}

impl ChunkStrategy {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Paragraph => "paragraph",
            Self::Sentence => "sentence",
            Self::Token => "token",
        }
    }

    #[allow(clippy::should_implement_trait)]
    pub fn from_str(s: &str) -> Option<Self> {
        match s.trim().to_ascii_lowercase().as_str() {
            "paragraph" => Some(Self::Paragraph),
            "sentence" => Some(Self::Sentence),
            "token" => Some(Self::Token),
            _ => None,
        }
    }
}

pub struct EmbeddingsConfig {
    pub provider: EmbeddingProvider,
    pub model: String,
    pub chunk_size: usize,
    pub chunk_overlap: usize,
    pub chunk_strategy: ChunkStrategy,
}

impl Default for EmbeddingsConfig {
    fn default() -> Self {
        Self {
            provider: EmbeddingProvider::Fireworks,
            model: "accounts/fireworks/models/qwen3-embedding-8b".to_string(),
            chunk_size: 8000,
            chunk_overlap: 10,
            chunk_strategy: ChunkStrategy::Paragraph,
        }
    }
}

pub struct EmbeddingResult {
    pub chunks: Vec<String>,
    pub embeddings: Vec<Vec<f64>>,
    pub model: String,
    pub provider: String,
}

pub fn chunk_text(
    text: &str,
    strategy: ChunkStrategy,
    chunk_size: usize,
    overlap_percent: usize,
) -> Vec<String> {
    match strategy {
        ChunkStrategy::Paragraph => chunk_by_paragraph(text, chunk_size, overlap_percent),
        ChunkStrategy::Sentence => chunk_by_sentence(text, chunk_size, overlap_percent),
        ChunkStrategy::Token => chunk_by_token(text, chunk_size, overlap_percent),
    }
}

fn chunk_by_paragraph(text: &str, chunk_size: usize, overlap_percent: usize) -> Vec<String> {
    let paragraphs: Vec<&str> = text
        .split("\n\n")
        .filter(|p| !p.trim().is_empty())
        .collect();
    let mut chunks = Vec::new();
    let mut current_chunk = String::new();
    let mut current_size = 0;

    for para in paragraphs {
        let para_size = para.len();
        if current_size + para_size > chunk_size && !current_chunk.is_empty() {
            chunks.push(current_chunk.trim().to_string());
            let _overlap_chars = (chunk_size * overlap_percent) / 100;
            current_chunk = String::new();
            current_size = 0;
        }

        if !current_chunk.is_empty() {
            current_chunk.push_str("\n\n");
            current_size += 2;
        }
        current_chunk.push_str(para);
        current_size += para_size;
    }

    if !current_chunk.trim().is_empty() {
        chunks.push(current_chunk.trim().to_string());
    }

    if chunks.is_empty() && !text.trim().is_empty() {
        chunks.push(text.trim().to_string());
    }

    chunks
}

fn chunk_by_sentence(text: &str, chunk_size: usize, overlap_percent: usize) -> Vec<String> {
    let sentence_enders = ['.', '!', '?', '¿', '¡'];
    let mut sentences: Vec<&str> = Vec::new();
    let mut current_start = 0;
    let bytes = text.as_bytes();

    for (i, &byte) in bytes.iter().enumerate() {
        let c = byte as char;
        if sentence_enders.contains(&c) && i + 1 < bytes.len() && bytes[i + 1].is_ascii_whitespace()
        {
            sentences.push(&text[current_start..=i]);
            current_start = i + 1;
            while current_start < bytes.len() && bytes[current_start].is_ascii_whitespace() {
                current_start += 1;
            }
        }
    }

    if current_start < text.len() {
        sentences.push(&text[current_start..]);
    }

    let mut chunks = Vec::new();
    let mut current_chunk = String::new();
    let mut current_size = 0;

    for sentence in sentences {
        let sentence_size = sentence.len();
        if current_size + sentence_size > chunk_size && !current_chunk.is_empty() {
            chunks.push(current_chunk.trim().to_string());
            let _overlap_chars = (chunk_size * overlap_percent) / 100;
            current_chunk = String::new();
            current_size = 0;
        }

        if !current_chunk.is_empty() {
            current_chunk.push(' ');
            current_size += 1;
        }
        current_chunk.push_str(sentence);
        current_size += sentence_size;
    }

    if !current_chunk.trim().is_empty() {
        chunks.push(current_chunk.trim().to_string());
    }

    if chunks.is_empty() && !text.trim().is_empty() {
        chunks.push(text.trim().to_string());
    }

    chunks
}

fn chunk_by_token(text: &str, chunk_size: usize, overlap_percent: usize) -> Vec<String> {
    let words: Vec<&str> = text.split_whitespace().collect();
    let mut chunks = Vec::new();
    let mut current_chunk = Vec::new();
    let mut current_size = 0;

    for word in words {
        let word_size = word
            .split(|c: char| c.is_ascii_punctuation())
            .filter(|s| !s.is_empty())
            .map(|s| s.chars().count())
            .sum::<usize>()
            .max(1);

        if current_size + word_size > chunk_size && !current_chunk.is_empty() {
            chunks.push(current_chunk.join(" "));
            let overlap_words = (chunk_size * overlap_percent) / 100 / 5;
            current_chunk = current_chunk[current_chunk
                .len()
                .saturating_sub(overlap_words.min(current_chunk.len()))..]
                .to_vec();
            current_size = current_chunk.iter().map(|w: &&str| w.len()).sum::<usize>();
        }

        current_chunk.push(word);
        current_size += word_size;
    }

    if !current_chunk.is_empty() {
        chunks.push(current_chunk.join(" "));
    }

    if chunks.is_empty() && !text.trim().is_empty() {
        chunks.push(text.trim().to_string());
    }

    chunks
}

pub fn embed_texts(
    config: &EmbeddingsConfig,
    texts: &[String],
) -> Result<EmbeddingResult, Box<dyn Error + Send + Sync>> {
    let all_chunks: Vec<String> = texts
        .iter()
        .flat_map(|text| {
            chunk_text(
                text,
                config.chunk_strategy,
                config.chunk_size,
                config.chunk_overlap,
            )
        })
        .collect();

    if all_chunks.is_empty() {
        return Ok(EmbeddingResult {
            chunks: vec![],
            embeddings: vec![],
            model: config.model.clone(),
            provider: config.provider.as_str().to_string(),
        });
    }

    let embeddings = embed_chunks(config, &all_chunks)?;

    Ok(EmbeddingResult {
        chunks: all_chunks,
        embeddings,
        model: config.model.clone(),
        provider: config.provider.as_str().to_string(),
    })
}

fn embed_chunks(
    config: &EmbeddingsConfig,
    chunks: &[String],
) -> Result<Vec<Vec<f64>>, Box<dyn Error + Send + Sync>> {
    match config.provider {
        EmbeddingProvider::Openai => embed_chunks_openai(&config.model, &config.api_key()?, chunks),
        EmbeddingProvider::Fireworks => {
            embed_chunks_fireworks(&config.model, &config.api_key()?, chunks)
        }
    }
}

pub fn embed_chunks_with_provider(
    provider: EmbeddingProvider,
    model: &str,
    chunks: &[String],
) -> Result<Vec<Vec<f64>>, Box<dyn Error + Send + Sync>> {
    let config = EmbeddingsConfig {
        provider,
        model: model.to_string(),
        ..EmbeddingsConfig::default()
    };
    embed_chunks(&config, chunks)
}

impl EmbeddingsConfig {
    fn api_key(&self) -> Result<String, Box<dyn Error + Send + Sync>> {
        let env_key = match self.provider {
            EmbeddingProvider::Openai => "OPENAI_API_KEY",
            EmbeddingProvider::Fireworks => "FIREWORKS_API_KEY",
        };
        env::var(env_key).map_err(|_| format!("{env_key} is not set in the environment").into())
    }

    #[allow(dead_code)]
    fn endpoint(&self) -> &'static str {
        match self.provider {
            EmbeddingProvider::Openai => "https://api.openai.com/v1/embeddings",
            EmbeddingProvider::Fireworks => "https://api.fireworks.ai/inference/v1/embeddings",
        }
    }
}

fn embed_chunks_fireworks(
    model: &str,
    api_key: &str,
    chunks: &[String],
) -> Result<Vec<Vec<f64>>, Box<dyn Error + Send + Sync>> {
    let client = Client::new();
    let base_url = "https://api.fireworks.ai/inference/v1/embeddings";

    let mut embeddings = Vec::with_capacity(chunks.len());

    for chunk in chunks {
        let payload = json!({
            "model": model,
            "input": chunk,
        });

        let response = client
            .post(base_url)
            .bearer_auth(api_key)
            .json(&payload)
            .send()?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().unwrap_or_default();
            return Err(format!("Fireworks API error {status}: {body}").into());
        }

        let body: serde_json::Value = response.json()?;
        let embedding = body["data"][0]["embedding"]
            .as_array()
            .ok_or("Missing embedding data from Fireworks API")?;

        let vector: Vec<f64> = embedding.iter().filter_map(|v| v.as_f64()).collect();

        embeddings.push(vector);
    }

    Ok(embeddings)
}

fn embed_chunks_openai(
    model: &str,
    api_key: &str,
    chunks: &[String],
) -> Result<Vec<Vec<f64>>, Box<dyn Error + Send + Sync>> {
    let client = Client::new();
    let base_url = "https://api.openai.com/v1/embeddings";

    let payload = json!({
        "model": model,
        "input": chunks,
    });

    let response = client
        .post(base_url)
        .bearer_auth(api_key)
        .json(&payload)
        .send()?;

    if !response.status().is_success() {
        let status = response.status();
        let body = response.text().unwrap_or_default();
        return Err(format!("OpenAI API error {status}: {body}").into());
    }

    let body: serde_json::Value = response.json()?;
    let data = body["data"]
        .as_array()
        .ok_or("Missing embedding data from OpenAI API")?;

    let mut embeddings: Vec<Vec<f64>> = Vec::with_capacity(data.len());

    for item in data {
        let embedding = item["embedding"]
            .as_array()
            .ok_or("Missing embedding array")?;

        let vector: Vec<f64> = embedding.iter().filter_map(|v| v.as_f64()).collect();

        embeddings.push(vector);
    }

    embeddings.sort_by_key(|e| e.len());

    Ok(embeddings)
}

File: mpipe/rchain/embeddings.py
# SPDX-License-Identifier: GPL-3.0-or-later
#
# mpipe - Multi-provider LLM CLI tools
# Copyright (C) 2026 Nicolas Schaetti and contributors
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

"""mpipe.rchain.embeddings module.

Notes
-----
This module is part of the Python port of the `mpipe` project.
"""
from __future__ import annotations

import os
import re
from dataclasses import dataclass
from enum import Enum

import requests


class EmbeddingProvider(str, Enum):
    """Embeddingprovider.

    Notes
    -----
    This class follows the same role and hierarchy as the Rust implementation.
    """
    OPENAI = "openai"
    FIREWORKS = "fireworks"

    def as_str(self) -> str:
        """As str.

        Parameters
        ----------
        None
            This callable does not accept explicit parameters.

        Returns
        -------
        str
            Returned value.
        """
        return self.value
    # end def as_str
# end class EmbeddingProvider


class ChunkStrategy(str, Enum):
    """Chunkstrategy.

    Notes
    -----
    This class follows the same role and hierarchy as the Rust implementation.
    """
    PARAGRAPH = "paragraph"
    SENTENCE = "sentence"
    TOKEN = "token"

    @classmethod
    def from_str(cls, value: str) -> "ChunkStrategy | None":
        """From str.

        Parameters
        ----------
        value : str
            Argument value.

        Returns
        -------
        'ChunkStrategy | None'
            Returned value.
        """
        raw = value.strip().lower()
        for strategy in cls:
            if strategy.value == raw:
                return strategy
            # end if
        # end for
        return None
    # end def from_str
# end class ChunkStrategy


@dataclass(slots=True)
class EmbeddingsConfig:
    """Embeddingsconfig.

    Notes
    -----
    This class follows the same role and hierarchy as the Rust implementation.
    """
    provider: EmbeddingProvider = EmbeddingProvider.FIREWORKS
    model: str = "accounts/fireworks/models/qwen3-embedding-8b"
    chunk_size: int = 8000
    chunk_overlap: int = 10
    chunk_strategy: ChunkStrategy = ChunkStrategy.PARAGRAPH

    def api_key(self) -> str:
        """Api key.

        Parameters
        ----------
        None
            This callable does not accept explicit parameters.

        Returns
        -------
        str
            Returned value.
        """
        env_key = "OPENAI_API_KEY" if self.provider == EmbeddingProvider.OPENAI else "FIREWORKS_API_KEY"
        value = os.getenv(env_key, "")
        if not value:
            raise ValueError(f"{env_key} is not set in the environment")
        # end if
        return value
    # end def api_key
# end class EmbeddingsConfig


@dataclass(slots=True)
class EmbeddingResult:
    """Embeddingresult.

    Notes
    -----
    This class follows the same role and hierarchy as the Rust implementation.
    """
    chunks: list[str]
    embeddings: list[list[float]]
    model: str
    provider: str
# end class EmbeddingResult


def chunk_text(text: str, strategy: ChunkStrategy, chunk_size: int, overlap_percent: int) -> list[str]:
    """Chunk text.

    Parameters
    ----------
    text : str
        Argument value.
    strategy : ChunkStrategy
        Argument value.
    chunk_size : int
        Argument value.
    overlap_percent : int
        Argument value.

    Returns
    -------
    list[str]
        Returned value.
    """
    if strategy == ChunkStrategy.PARAGRAPH:
        return _chunk_by_paragraph(text, chunk_size, overlap_percent)
    # end if
    if strategy == ChunkStrategy.SENTENCE:
        return _chunk_by_sentence(text, chunk_size, overlap_percent)
    # end if
    return _chunk_by_token(text, chunk_size, overlap_percent)
# end def chunk_text


def embed_texts(config: EmbeddingsConfig, texts: list[str]) -> EmbeddingResult:
    """Embed texts.

    Parameters
    ----------
    config : EmbeddingsConfig
        Argument value.
    texts : list[str]
        Argument value.

    Returns
    -------
    EmbeddingResult
        Returned value.
    """
    all_chunks: list[str] = []
    for text in texts:
        all_chunks.extend(
            chunk_text(text, config.chunk_strategy, config.chunk_size, config.chunk_overlap)
        )
    # end for

    if not all_chunks:
        return EmbeddingResult(
            chunks=[],
            embeddings=[],
            model=config.model,
            provider=config.provider.as_str(),
        )
    # end if

    embeddings = embed_chunks(config, all_chunks)
    return EmbeddingResult(
        chunks=all_chunks,
        embeddings=embeddings,
        model=config.model,
        provider=config.provider.as_str(),
    )
# end def embed_texts


def embed_chunks(config: EmbeddingsConfig, chunks: list[str]) -> list[list[float]]:
    """Embed chunks.

    Parameters
    ----------
    config : EmbeddingsConfig
        Argument value.
    chunks : list[str]
        Argument value.

    Returns
    -------
    list[list[float]]
        Returned value.
    """
    if config.provider == EmbeddingProvider.OPENAI:
        return _embed_chunks_openai(config.model, config.api_key(), chunks)
    # end if
    return _embed_chunks_fireworks(config.model, config.api_key(), chunks)
# end def embed_chunks


def embed_chunks_with_provider(
    provider: EmbeddingProvider,
    model: str,
    chunks: list[str],
) -> list[list[float]]:
    """Embed chunks with provider.

    Parameters
    ----------
    provider : EmbeddingProvider
        Argument value.
    model : str
        Argument value.
    chunks : list[str]
        Argument value.

    Returns
    -------
    list[list[float]]
        Returned value.
    """
    config = EmbeddingsConfig(provider=provider, model=model)
    return embed_chunks(config, chunks)
# end def embed_chunks_with_provider


def _chunk_by_paragraph(text: str, chunk_size: int, overlap_percent: int) -> list[str]:
    """ chunk by paragraph.

    Parameters
    ----------
    text : str
        Argument value.
    chunk_size : int
        Argument value.
    overlap_percent : int
        Argument value.

    Returns
    -------
    list[str]
        Returned value.
    """
    del overlap_percent
    paragraphs = [p for p in text.split("\n\n") if p.strip()]
    chunks: list[str] = []
    current: list[str] = []
    current_size = 0
    for paragraph in paragraphs:
        paragraph_size = len(paragraph)
        if current and current_size + paragraph_size > chunk_size:
            chunks.append("\n\n".join(current).strip())
            current = []
            current_size = 0
        # end if
        current.append(paragraph)
        current_size += paragraph_size + (2 if current_size > 0 else 0)
    # end for
    if current:
        chunks.append("\n\n".join(current).strip())
    # end if
    if not chunks and text.strip():
        chunks.append(text.strip())
    # end if
    return chunks
# end def _chunk_by_paragraph


def _chunk_by_sentence(text: str, chunk_size: int, overlap_percent: int) -> list[str]:
    """ chunk by sentence.

    Parameters
    ----------
    text : str
        Argument value.
    chunk_size : int
        Argument value.
    overlap_percent : int
        Argument value.

    Returns
    -------
    list[str]
        Returned value.
    """
    del overlap_percent
    sentences = [s.strip() for s in re.split(r"(?<=[.!?¿¡])\s+", text) if s.strip()]
    chunks: list[str] = []
    current: list[str] = []
    current_size = 0
    for sentence in sentences:
        sentence_size = len(sentence)
        if current and current_size + sentence_size > chunk_size:
            chunks.append(" ".join(current).strip())
            current = []
            current_size = 0
        # end if
        current.append(sentence)
        current_size += sentence_size + (1 if current_size > 0 else 0)
    # end for
    if current:
        chunks.append(" ".join(current).strip())
    # end if
    if not chunks and text.strip():
        chunks.append(text.strip())
    # end if
    return chunks
# end def _chunk_by_sentence


def _chunk_by_token(text: str, chunk_size: int, overlap_percent: int) -> list[str]:
    """ chunk by token.

    Parameters
    ----------
    text : str
        Argument value.
    chunk_size : int
        Argument value.
    overlap_percent : int
        Argument value.

    Returns
    -------
    list[str]
        Returned value.
    """
    words = text.split()
    chunks: list[str] = []
    current: list[str] = []
    current_size = 0
    for word in words:
        word_size = max(1, len(re.sub(r"\W+", "", word)))
        if current and current_size + word_size > chunk_size:
            chunks.append(" ".join(current))
            overlap_words = (chunk_size * overlap_percent) // 100 // 5
            current = current[max(0, len(current) - min(overlap_words, len(current))) :]
            current_size = sum(len(item) for item in current)
        # end if
        current.append(word)
        current_size += word_size
    # end for
    if current:
        chunks.append(" ".join(current))
    # end if
    if not chunks and text.strip():
        chunks.append(text.strip())
    # end if
    return chunks
# end def _chunk_by_token


def _embed_chunks_fireworks(model: str, api_key: str, chunks: list[str]) -> list[list[float]]:
    """ embed chunks fireworks.

    Parameters
    ----------
    model : str
        Argument value.
    api_key : str
        Argument value.
    chunks : list[str]
        Argument value.

    Returns
    -------
    list[list[float]]
        Returned value.
    """
    url = "https://api.fireworks.ai/inference/v1/embeddings"
    session = requests.Session()
    vectors: list[list[float]] = []
    for chunk in chunks:
        response = session.post(
            url,
            headers={"Authorization": f"Bearer {api_key}"},
            json={"model": model, "input": chunk},
            timeout=60,
        )
        if not response.ok:
            raise RuntimeError(f"Fireworks API error {response.status_code}: {response.text}")
        # end if
        body = response.json()
        data = body.get("data")
        if not isinstance(data, list) or not data:
            raise RuntimeError("Missing embedding data from Fireworks API")
        # end if
        embedding = data[0].get("embedding")
        if not isinstance(embedding, list):
            raise RuntimeError("Missing embedding data from Fireworks API")
        # end if
        vectors.append([float(value) for value in embedding])
    # end for
    return vectors
# end def _embed_chunks_fireworks


def _embed_chunks_openai(model: str, api_key: str, chunks: list[str]) -> list[list[float]]:
    """ embed chunks openai.

    Parameters
    ----------
    model : str
        Argument value.
    api_key : str
        Argument value.
    chunks : list[str]
        Argument value.

    Returns
    -------
    list[list[float]]
        Returned value.
    """
    url = "https://api.openai.com/v1/embeddings"
    response = requests.post(
        url,
        headers={"Authorization": f"Bearer {api_key}"},
        json={"model": model, "input": chunks},
        timeout=60,
    )
    if not response.ok:
        raise RuntimeError(f"OpenAI API error {response.status_code}: {response.text}")
    # end if
    body = response.json()
    data = body.get("data")
    if not isinstance(data, list):
        raise RuntimeError("Missing embedding data from OpenAI API")
    # end if
    vectors: list[list[float]] = []
    for item in data:
        embedding = item.get("embedding") if isinstance(item, dict) else None
        if not isinstance(embedding, list):
            raise RuntimeError("Missing embedding array")
        # end if
        vectors.append([float(value) for value in embedding])
    # end for
    vectors.sort(key=len)
    return vectors
# end def _embed_chunks_openai
